File: android-client/app/src/main/java/com/example/android_client/content/ContentPluginRegistry.kt
package com.example.android_client.content

/**
 * Registry holding all registered content plugins, keyed by contentType.
 */
class ContentPluginRegistry {
    private val plugins = mutableMapOf<String, ContentPlugin>()

    fun register(plugin: ContentPlugin) {
        plugins[plugin.contentType] = plugin
    }

    fun get(contentType: String): ContentPlugin? = plugins[contentType]

    fun getAll(): Collection<ContentPlugin> = plugins.values

    fun contentTypes(): Set<String> = plugins.keys
}

File: windows-client/src/Hikari.WindowsClient/Content/ContentPluginRegistry.cs
namespace Hikari.WindowsClient.Content;

/// <summary>
/// Registry holding all registered content plugins, keyed by content type.
/// Mirrors <c>android-client/app/src/content/ContentPluginRegistry.kt</c>.
/// </summary>
public sealed class ContentPluginRegistry
{
    private readonly Dictionary<string, IContentPlugin> _plugins = new(StringComparer.OrdinalIgnoreCase);
    private readonly List<IContentPlugin> _ordered = new();

    public void Register(IContentPlugin plugin)
    {
        if (_plugins.TryAdd(plugin.ContentType, plugin))
        {
            _ordered.Add(plugin);
        }
        else
        {
            _plugins[plugin.ContentType] = plugin;
            var index = _ordered.FindIndex(p =>
                string.Equals(p.ContentType, plugin.ContentType, StringComparison.OrdinalIgnoreCase));
            if (index >= 0) _ordered[index] = plugin;
        }
    }

    public IContentPlugin? Get(string contentType) =>
        _plugins.TryGetValue(contentType, out var plugin) ? plugin : null;

    /// <summary>All plugins, in registration order.</summary>
    public IReadOnlyList<IContentPlugin> GetAll() => _ordered;

    public IReadOnlyCollection<string> ContentTypes() => _plugins.Keys;
}
